### PR TITLE
Continuous h1-pool prefetching

### PR DESCRIPTION
**Pull request by oschaaf**
Now that we derive our own H1 pool implementation, clean up the way we implement connection prefetching, as well as make it a continuous process instead of only doing it up front.

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by oschaaf** [comment]
@envoyproxy/nighthawk-reviewers this PR is ready for review

File: test/benchmark_http_client_test.cc [hunk 1]
```diff
@@ -220,33 +220,6 @@ TEST_F(BenchmarkClientHttpTest, StatusTrackingInOnComplete) {
   client_.reset();
 }
 
-TEST_F(BenchmarkClientHttpTest, ConnectionPrefetching) {
```
Comment (mum4k) [inline test/benchmark_http_client_test.cc]: Do we have any new behaviors that we should test here? Say the difference between having the setPrefetchConnections enabled and disabled?

Comment (oschaaf) [reply test/benchmark_http_client_test.cc]: Sorry, should have probably mentioned this in the PR description, but the test at https://github.com/envoyproxy/nighthawk/blob/master/test/integration/test_integration_basics.py#L246 is also in place.

That python-based integration testing is often convenient for end-to-end tests, but was not around initially. With the changes made here, this C++ test needed maintenance, but I figured it didn't really add any value on top of the python-based test. So I dropped it instead.

Comment (mum4k) [reply test/benchmark_http_client_test.cc]: Thank you @oschaaf.

File: source/client/benchmark_client_impl.cc [hunk 2]
```diff
@@ -21,11 +21,18 @@ using namespace std::chrono_literals;
...
-void Http1PoolImpl::createConnections(const uint32_t connection_limit) {
-  ENVOY_LOG(error, "Prefetching {} connections.", connection_limit);
-  for (uint32_t i = 0; i < connection_limit; i++) {
-    createNewConnection();
+Envoy::Http::ConnectionPool::Cancellable*
+Http1PoolImpl::newStream(Envoy::Http::StreamDecoder& response_decoder,
+                         Envoy::Http::ConnectionPool::Callbacks& callbacks) {
+  // In prefetch mode we try to keep the amount of connections at the configured limit.
+  if (prefetch_connections_) {
+    while (host_->cluster().resourceManager(priority_).connections().canCreate()) {
```
Comment (htuch) [inline source/client/benchmark_client_impl.cc]: What guarantee do we have that this doesn't live lock?

Comment (oschaaf) [reply source/client/benchmark_client_impl.cc]: So.. how would you see this lock? 
Note: each worker in NH has a unique cluster assigned to it / they are not shared across workers. As far as I can tell, the vanilla Envoy pool implementations also use this in vanilla Envoy.

Comment (htuch) [reply source/client/benchmark_client_impl.cc]: I'm wondering what guarntees we have that `canCreate()` returns false at some point?

Comment (oschaaf) [reply source/client/benchmark_client_impl.cc]: The call to `ConnPoolImpl::createNewConnection()` unconditionally constructs a new `ActiveClient`, which in turn unconditionally and directly calls `resourceManager(parent_.priority_).connections().inc();`.

The resource manager is configured to cap the number of connections to our `--connection` limit option, and at some point we're going to hit that at which point `canCreate()` will return false.

**Comment by htuch** [comment]
@oschaaf needs master merge.

**Comment by oschaaf** [comment]
> @oschaaf needs master merge.

@htuch done!

**Pull request merged**